In [1]:
# pip install opencv-python
# pip install pytesseract
import cv2
import pytesseract

In [2]:
# set path of tesseract-ocr in windows it is not needed in linux
pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'

In [3]:
moon_img=cv2.imread('img1_moonshine.jpg')
moon_img

array([[[ 6,  5,  1],
        [ 6,  5,  1],
        [ 6,  5,  1],
        ...,
        [ 9,  4,  1],
        [ 9,  4,  1],
        [ 9,  4,  1]],

       [[ 6,  5,  1],
        [ 6,  5,  1],
        [ 6,  5,  1],
        ...,
        [ 9,  4,  1],
        [ 9,  4,  1],
        [ 9,  4,  1]],

       [[ 6,  5,  1],
        [ 6,  5,  1],
        [ 6,  5,  1],
        ...,
        [ 9,  4,  1],
        [ 9,  4,  1],
        [ 9,  4,  1]],

       ...,

       [[ 8,  3,  0],
        [ 8,  3,  0],
        [ 9,  4,  1],
        ...,
        [12,  7,  4],
        [12,  7,  4],
        [12,  7,  4]],

       [[ 8,  3,  0],
        [ 8,  3,  0],
        [ 9,  4,  1],
        ...,
        [12,  7,  4],
        [12,  7,  4],
        [12,  7,  4]],

       [[ 8,  3,  0],
        [ 8,  3,  0],
        [ 9,  4,  1],
        ...,
        [12,  7,  4],
        [12,  7,  4],
        [12,  7,  4]]], shape=(173, 291, 3), dtype=uint8)

In [4]:
# Read the image
cv2.imshow("Moonshine",moon_img)
cv2.waitKey()
cv2.destroyAllWindows()

In [5]:
# Extracting the text from the image
text = pytesseract.image_to_string(moon_img)
print(text)
# There is two texts but only moonshine is read. That is bcz the clarity of other text is less. 
# so inorder to extract that we can convert the image to gray scale

MoonShine



In [6]:
# To get the pixels ie coordinates (complete data) instead of just getting the text
# we use image_to_data function
# The o/p will be a dictionary
details = pytesseract.image_to_data(moon_img,output_type=pytesseract.Output.DICT)
details

{'level': [1, 2, 3, 4, 5],
 'page_num': [1, 1, 1, 1, 1],
 'block_num': [0, 1, 1, 1, 1],
 'par_num': [0, 0, 1, 1, 1],
 'line_num': [0, 0, 0, 1, 1],
 'word_num': [0, 0, 0, 0, 1],
 'left': [0, 43, 43, 43, 43],
 'top': [0, 75, 75, 75, 75],
 'width': [291, 210, 210, 210, 210],
 'height': [173, 32, 32, 32, 32],
 'conf': [-1, -1, -1, -1, 91],
 'text': ['', '', '', '', 'MoonShine']}

{'level': [1, 2, 3, 4, 5],

 'page_num': [1, 1, 1, 1, 1],

 'block_num': [0, 1, 1, 1, 1],

 'par_num': [0, 0, 1, 1, 1],

 'line_num': [0, 0, 0, 1, 1],

 'word_num': [0, 0, 0, 0, 1],

 'left': [0, 43, 43, 43, 43],                   # the text is at the end so we need to consider the last number here

 'top': [0, 75, 75, 75, 75],

 'width': [291, 210, 210, 210, 210],

 'height': [173, 32, 32, 32, 32],

 'conf': [-1, -1, -1, -1, 91],                 # confidentiality : confidentiality number of space is -1, 

                                                special character 0,1 etc ie small numbers, 

                                                >50 confidentiality is readed as text (ie 50+ is only considered as text)

 'text': ['', '', '', '', 'MoonShine']}        # the space before moonshine is marked as space

 Left, Top        => (43,75)   => (x1,y1)
 
 width, height    => (210,32)  => (w,h)

In [7]:
#  To draw box
x,y,w,h = 43,75,210,32
cv2.rectangle(moon_img,(x,y),(x+w,y+h),(25,0,255),4) # in this color is in BGR
cv2.imshow("Moonshine",moon_img)
cv2.waitKey()
cv2.destroyAllWindows()

In [8]:
para_img = cv2.imread('img3_para.jpg')
para_img

array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       ...,

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]

In [9]:
cv2.imshow("Para",para_img)
cv2.waitKey()
cv2.destroyAllWindows()

In [10]:
text_para = pytesseract.image_to_string(para_img)
print(text_para)

This is a lot of 12 point text to test the
ocr code and see if it works on all types
of file format.

The quick brown dog jumped over the
lazy fox. The quick brown dog jumped
over the lazy fox. The quick brown dog
jumped over the lazy fox. The quick
brown dog jumped over the lazy fox.



In [11]:
details_para = pytesseract.image_to_data(para_img,output_type=pytesseract.Output.DICT)
details_para

{'level': [1,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5],
 'page_num': [1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1],
 'block_num': [0,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,

In [12]:
# details_para['text']

In [13]:
# To get the starting and ending position calculate the dictionary length
# starting will be at 0
text_para_len = len(details_para['text'])
text_para_len

# ending will be at 72 and it is in text_para_len

72

In [14]:
for i in range(text_para_len):
    if details_para['conf'][i] > 50: # To avoid drawing unwanted boxes for spaces etc, and to draw only for text
        x = details_para['left'][i]
        y = details_para['top'][i]
        w = details_para['width'][i]
        h = details_para['height'][i]
        cv2.rectangle(para_img,(x,y),(x+w,y+h),(0,255,0),2)
cv2.imshow("Para_img",para_img)
cv2.waitKey()
cv2.destroyAllWindows()